web-module/src/main/webapp/resources/structEditJS/autocomplete.mjs

```diff
@@ -139,35 +139,95 @@ class CelAutocompleteInitialiser {
+    const theAddNewPopup = window.open(addNewUrl, '_blank', 'popup=true');
+    theAddNewPopup.addEventListener('message', (ev) => this.#handleAddNewMessage(ev, selectElem));
+  }
+
+  #addNewButtonElem(selectElem, addNewUrl) {
+    const classField = selectElem.dataset.classField || ''; 
+    const type = selectElem.dataset.autocompleteType || '';
+    const cellRef = selectElem.dataset.cellRef || '';
+    const buttonText = window.celMessages.structEditor.autocomplete['addNewButtonText_' + cellRef]
+                || window.celMessages.structEditor.autocomplete['addNewButtonText_' + classField]
+                || window.celMessages.structEditor.autocomplete['addNewButtonText_' + type]
+                || window.celMessages.structEditor.autocomplete['addNewButtonText_default']
+                || 'nothing found? add new';
+    const iconElem = document.createElement('span');
+    iconElem.classList.add('halflings', 'halflings-plus-sign');
+    const buttonInnerElem = document.createElement('div');
+    buttonInnerElem.appendChild(iconElem);
+    buttonInnerElem.classList.add('view_cel_buttonLink');
+    buttonInnerElem.setAttribute('data-url', addNewUrl);
+    buttonInnerElem.insertAdjacentText('beforeend', buttonText)
+    buttonInnerElem.addEventListener('click', 
+        (ev) => this.#addNewButtonClickHandler(ev, selectElem));
+    const buttonElem = document.createElement('div');
+    buttonElem.appendChild(buttonInnerElem);
+    buttonElem.classList.add('box', 'cel_button', 'struct_autocomplete_addnew');
+    const itemElem = document.createElement('div');
+    itemElem.appendChild(buttonElem);
+    itemElem.classList.add('result', 'clearfix');
+    return itemElem;
+  }
+
+  #concatAddNewButtonToResults(selectElem, response) {
+    const results = response.results || [];
+    const addNewUrl = response.addNewUrl || '';
+    if (!response.hasMore && (addNewUrl !== '')) {
+      console.debug('add new url', addNewUrl);
+      results.push({
+          'html' : this.#addNewButtonElem(selectElem, addNewUrl)
+      });
+    }
+    console.debug('concatAddNewButtonToResults return', results);
+    return results;
+  }
+
   /**
    * parse the results into the format expected by Select2
    * since we are using custom formatting functions we do not need to
    * alter the remote JSON data, except to indicate that infinite
    * scrolling can be used
    */
-  processResultsFunc(response, params) {
+  processResultsFunc(selectElem, response, params) {
     params.page = params.page || 1;
+    console.debug('processResultsFunc', selectElem, response, params);
     return {
-      results: (response.results || [])
-        .map(elem => {
+      results: this.#concatAddNewButtonToResults(selectElem, response).map(elem => {
           elem.id = elem.fullName;
           elem.text = elem.name;
           return elem;
-        }).filter(elem => elem.id && elem.text),
+        }).filter(elem => elem.html || elem.id && elem.text),
```

Comment:
I think keeping the previous version but concating the button to the array is simpler and easier to understand:
```
      results: (response.results || [])
        .map(elem => {
          elem.id = elem.fullName;
          elem.text = elem.name;
          return elem;
        }).filter(elem => elem.id && elem.text)
        .concat([this.#createAddNewButton(response.addNewUrl, selectElem)].filter(Boolean)),
```
`#createAddNewButton` simply needs to return the element or undefined.

Reply:
It depends on the personal preference over functional/fluent programming and imperativ programming. Yet, for you I surly change it :-) Nevertheless, the extension of the filter with `elem.html ||` is needed to allow html beeing possibly generated serverside.